### PR TITLE
executor: Introduce a new timeout executor

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

The new executor uses a timeout to determine if the command it is
instructed to execute should run or not.

There are 3 scenarios that are supported:
- Run successful command before timeout is reached.
- Run failing command before timeout is reached.
- Silently skip running command after timeout is reached.

An error is expected to be returned only for the 2nd scenario above.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/executor/pool.go [hunk 1]
```diff
@@ -27,29 +27,21 @@ import (
...
 // Otherwise, it will create new RateLimitedExecutor with a new underlying rate-limiter, store and return it.
 func (c *RateLimitedExecutorPool) LoadOrStore(key interface{}) *RateLimitedExecutor {
-	element, exists := c.pool.Load(key)
-	if !exists {
-		rateLimiter := c.rateLimiterCreator.New()
-		element, _ := c.pool.LoadOrStore(key, NewRateLimitedExecutor(&rateLimiter))
-		return element.(*RateLimitedExecutor)
-	}
+	rateLimit := c.creator.New()
+	element, _ := c.Map.LoadOrStore(key, NewRateLimitedExecutor(&rateLimit))
```
Comment (RamLavi) [inline pkg/executor/pool.go:45]: I like the embedding trick (never seen it before), but why is this change relevant to the timeout executor? I would move this commit to another PR

File: pkg/executor/timeout/timeoutpool.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func NewTimeoutExecutorPool(creator TimerCreator) *ExecutorPool {
+	return &ExecutorPool{
+		Map:     sync.Map{},
+		creator: creator,
+	}
+}
+
+// LoadOrStore returns the existing RateLimitedExecutor for the key if present.
```
Comment (RamLavi) [inline pkg/executor/timeout/timeoutpool.go:38]: copied comment, please change:
RateLimitedExecutor --> Executor
with a new underlying rate-limiter --> with a new underlying timer

**Review by RamLavi — CHANGES_REQUESTED** [review]
Like! 
I think the 1st commit should be in a separate PR ( it has nothing to do with the new executer)
Added small comments

File: pkg/executor/pool.go (hunk 1 above)
Comment (EdDev) [reply pkg/executor/pool.go:45]: Because it allowed me to copy a simpler version for the new executor type.

Now that it is already working, I guess I can extract this out to a separate PR.

Comment (EdDev) [reply pkg/executor/pool.go:45]: Done.
Move to #8177

File: pkg/executor/timeout/timeoutpool.go (hunk 2 above)
Comment (EdDev) [reply pkg/executor/timeout/timeoutpool.go:38]: Done

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @EdDev.
Please see the inline comments.

File: pkg/executor/timeout/timeoutpool.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	return &ExecutorPool{
+		Map:     sync.Map{},
+		creator: creator,
+	}
+}
+
+// LoadOrStore returns the existing executor for the key if present.
+// Otherwise, it will create a new timeout executor, store and return it.
+func (c *ExecutorPool) LoadOrStore(key interface{}) *Executor {
+	timeout := c.creator.New()
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timeoutpool.go:41]: Done
I could not use `timer` because the editor complains it is shadowing the interface name.

File: pkg/executor/timeout/timeoutexecutor.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+type timer interface {
+	Expired() bool
+}
+
+type Executor struct {
+	timer timer
+}
+
+func NewExecutor(timer timer) *Executor {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timeoutexecutor.go:30]: I think you are right, we do not need pointers at all now. This executor is immutable.

File: pkg/executor/timeout/timeoutexecutor.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	return &Executor{timer: timer}
+}
+
+// Exec will execute the given func when the underlying timer has not expired.
+// In case the timer expired, the command is not executed and nil is returned.
+func (t Executor) Exec(command func() error) error {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timeoutexecutor.go:36]: I think that in most cases, this will not be needed.
The caller should not care if it was called or not.

The caller needs to decide if a command should be called or not.
Once a decision was made to call it, this executor will either execute the command successfully or not call it at all. The caller does not care.

If the caller cares, we either need to extend this with a `IsTimeout` or with a special error returned. But I do not think we need it now.

Either way, we should wait for @RamLavi feedback on using it in that PR scenario.

File: pkg/executor/timeout/timer.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package timeout
+
+import (
+	"time"
+)
+
+type Timer struct {
+	timeoutTime time.Time
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timer.go:27]: Done

File: pkg/executor/timeout/timer.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func NewTimer(timeout time.Duration) Timer {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timer.go:30]: Done

File: pkg/executor/timeout/timer.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+type TimerCreator struct {
+	timeout time.Duration
+}
+
+func NewTimerCreator(timeout time.Duration) TimerCreator {
+	return TimerCreator{timeout: timeout}
+}
+
+func (t *TimerCreator) New() Timer {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/executor/timeout/timer.go:48]: Done

**Comment by orelmisan** [comment]
Thank you for the changes @EdDev.

File: pkg/executor/timeout/timeoutexecutor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/executor/timeout/timeoutexecutor.go:36]: I agree with @EdDev. Currently we don't need to know if the timeout has passed. If the need for it ever arises, we can extend the API without issues.

**Review by RamLavi — APPROVED** [review]
/lgtm

**Review by orelmisan — APPROVED** [review]

**Comment by EdDev** [comment]
We should wait and see if it is indeed relevant for usage before proceeding.

/hold

**Comment by EdDev** [comment]
At the moment, this does not seem to be needed.
Closing for the time being.
